start path lesson (#71)
carpentries-incubator/medical-image-processing: episodes/pathology.md

 
 
 Tiled images are made in a pyramidal structure, and at each level of the pyramid after zero there is a lower resolution downscaled version of the image. Each level of the pyramid has tiles, smaller parts of the image that are quilted into the whole image. Creating images in this way allows much more efficient access and management of images because you can load and display images at only one level. 
 An OME-TIFF file can be one or more tiled TIFFs and some XML in the header. An SVS file is essentially a tiled TIFF image that has a few additional things like overview image. Various types of pathology files can be somehow packed into a DICOM which has it's own file specification. 

## Libraries

There is actually very little in the way of open pure python libraries that deal with pathology images. 
A promising new contender is [TIA toolbox](https://tia-toolbox.readthedocs.io/en/latest/) ; however this is a relatively new library. 
There are better known (in the small community of researchers in digital pathology) more popular libraries that have binders or other tricks that allow you to process pathology images in Python. 

> Probably the most popular is openslide, but other contenders include [Bio-Formats](http://www.openmicroscopy.org/bio-formats/) (which is both a libraru and a tool) and QuPath which has the distinct advantage of being able to work with OME-TIFF directly. 

Reviewer: QuPath is mainly a UI tool, that relies on libraries (Bioformats/OpenSlide) as "backends" to load the slide. It's different in that sense—it can be used directly as a tool for viewing, unlike BioFormats/OpenSlide, which is code only.

Author: Ok, adressed. Thanks for all your review here and elsewhere.

---
title: "Working with Pathology Images"
teaching: 60
exercises: 30
---

:::::::::::::::::::::::::::::::::::::: questions 

- What are some open libraries can be used with pathology?
- How are pathology slides different from a photo?
- What is the structure of a pathology slide?

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: objectives

- Describe some open source tools and libraries for pathology images
- Discuss  common file formats for pathology slides
- Explain tiling and pyramidal file formats
- Show basic code in openslide

::::::::::::::::::::::::::::::::::::::::::::::::

## Introduction

Pathologists deal with multiple types of images. Some pathologists may use photography when recording gross (or macroscopic) pathology data. The processing of digital photography is a straightforward application of general image processing. However the workhorse type image of pathology is the histopathological image, and this is very different from a photograph. 

> In the below picture you can see an image with gross and histopathology of signet ring cell carcinoma metastasis to the ovary. The histopathology images are stained microscopy image. The gross pathology images are photographs.

Reviewer 2: maybe it could be explained what signet ring cell carcinoma means...

Author: There is a reason I picked this cancer, but we want to not get too side tracked into medical stuff. But I will definitely I will specify which are photos and which are histo images. I forget that this is not obvious.

<img src="fig/path_metastasis.jpg" alt="Gross and histopathology" width="70%;"/>

*Nakamura, Yoshiaki; Hiramatsu, Ayako; Koyama, Takafumi; Oyama, Yu; Tanaka, Ayuko; Honma, Koichi, CC BY 3.0 <https://creativecommons.org/licenses/by/3.0>, via Wikimedia Commons*

In this lesson we will cover how to work with histopathology images as they differ from simpler multi-channel color image like a modern digital camera often produces. 

## File formats

In terms of histopathological images, you may find them in several file formats. Not every library or tool for such imaging works with every format. Below is a table with several file formats.

| Format Name |Extension|Notes|  More info|
| ----------- |---------|-----|-----------|
| TIFF     |.tiff|tag image file format| can be tiled |https://www.loc.gov/preservation/digital/formats/fdd/fdd000022.shtml|
| OME-TIFF       |.ome.tiff, ome.tif, .ome.tf2, .ome.tf8 or .ome.btf | includes XML and may be multiple TIFFs|https://brainder.org/2012/09/23/the-nifti-file-format/|
| SVS         | .svs    | Scanned Virtual Slide  |https://www.dicomstandard.org/  |

> | DICOM       |.dcm or none|supports an included WSI file|https://dicom.nema.org/dicom/dicomwsi/|

Reviewer 2: none if part of a DICOM media..

Author: I just meant there are a lot of dicom files that just have no file extension on them (at least no visible file extension and it drives me crazy).

Reviewer 2: yes, this is clear.

 The table above is far from complete because any vendors have their own file format, however the community is consolidating towards shared file formats that are not vendor specific. The table is organized in a manner of ascending complexity. A simple general TIFF file is very close to a simple raster image. Such a file for a high resolution image would be quite large and potentially hard to manage based on size alone. Tiling in a way to more efficiently manage huge images. 

>  Tiling  approaches will be very familiar to anyone who comes from the world of GIS, mapping or digital geography. Maps provide a really obvious way to think about the advantages of pyramidal tiling. Imagine if every time you wanted to get a map with directions, and check out a picture of your destination, you needed to load a detailed map picture of the entire world, literally. It would be incredibly inefficient and taxing on your computer or smartphone. It's much more efficient to just load what you need. There are many ways this can be done but practically some variation of pyramidal tiling is usually used. The same is true with tiled pathology images.

Reviewer 2: Are all tiling approaches are pyramidal?

Author: No. I'll clarify this.

Reviewer 2: ok thanks, i will review this when ready!

Author: New version addresses on lines 54/55

>  Tiled images are made in a pyramidal structure, and at each level of the pyramid after zero there is a lower resolution downscaled version of the image. Each level of the pyramid has tiles, smaller parts of the image that are quilted into the whole image. Creating images in this way allows much more efficient access and management of images because you can load and display images at only one level. 

Reviewer 2: > Tiled images are made in a pyramidal structure

Maybe say this earlier...

Author: Changed in new version

>  An OME-TIFF file can be one or more tiled TIFFs and some XML in the header. An SVS file is essentially a tiled TIFF image that has a few additional things like overview image. Various types of pathology files can be somehow packed into a DICOM which has it's own file specification. 

Reviewer 2: This is nice..I would only explain this a little better:
> Various types of pathology files can be somehow packed into a DICOM which has it's own file specification.

Author: I'm adding an extra sentence: "DICOMs could be conceptualized as a containers which can contain many sorts of imaging data, e.g. JPEGS of X-rays, including OME-TIFFs or other pathology file types. "

Reviewer 2: This is nice!

## Libraries

> There is actually very little in the way of open pure python libraries that deal with pathology images. 

Reviewer 2: ```suggestion
There is actually very little in the way of open pure Python libraries that deal with pathology images. 
```

Author: fixed

> A promising new contender is [TIA toolbox](https://tia-toolbox.readthedocs.io/en/latest/) ; however this is a relatively new library. 

Reviewer 2: > A promising new contender is [TIA toolbox](https://tia-toolbox.readthedocs.io/en/latest/) ; however this is a relatively new library.

Their software looks quite comprehensive with several versions...

Author: Yes, but experts I spoke to see it as a new kid- not the main tool kits everyone knows.

There are better known (in the small community of researchers in digital pathology) more popular libraries that have binders or other tricks that allow you to process pathology images in Python. 

> Probably the most popular is openslide, but other contenders include [Bio-Formats](http://www.openmicroscopy.org/bio-formats/) (which is both a libraru and a tool) and QuPath which has the distinct advantage of being able to work with OME-TIFF directly. 

Reviewer 2: Add a link to openslide and QuPath

> Probably the most popular is openslide, but other contenders include [Bio-Formats](http://www.openmicroscopy.org/bio-formats/) (which is both a libraru and a tool) and QuPath which has the distinct advantage of being able to work with OME-TIFF directly. 

Reviewer 2: ```suggestion
Probably the most popular is [openslide]( (https://openslide.org/)), but other contenders include [Bio-Formats](http://www.openmicroscopy.org/bio-formats/) (which is both a library and a tool) and QuPath which has the distinct advantage of being able to work with OME-TIFF directly. 
```
QuPath also needs a link.

QuPath is mainly a UI tool, that relies on libraries (Bioformats/OpenSlide) as "backends" to load the slide. It can be used directly as a tool for viewing, unlike BioFormats/OpenSlide, which are code only.

> One potential disadvantage of QuPath if you code with it is that usually requires some Java knowledge. Bio-Formats has Python bindings but is also mainly a Java tool which can also be used as a Java or C++ library. To keep things simple we will use openslide as the vehicle to explore histopathology images in this lesson.

Reviewer 2: ```suggestion
One potential disadvantage of QuPath is that it usually requires some Java knowledge. Bio-Formats has Python bindings but is also mainly a Java tool which can also be used as a Java or C++ library. To keep things simple we will use openslide as the vehicle to explore histopathology images in this lesson.
```

Author: Better clarified, by changing in a different way

#### Reading TIFF histopathology Images

> [Openslide](https://openslide.org/) is actually written in C, but you can also use Java or Python to work with is due to bindings.To run the code you will need to work in the `pathy` environment created with the environment_pathology.yml file.

Reviewer 2: ```suggestion
`environment_pathology.yml` file.
```

> [Openslide](https://openslide.org/) is actually written in C, but you can also use Java or Python to work with is due to bindings.To run the code you will need to work in the `pathy` environment created with the environment_pathology.yml file.

Author: Met you half way, but I want the link

> [Openslide](https://openslide.org/) is actually written in C, but you can also use Java or Python to work with is due to bindings.To run the code you will need to work in the `pathy` environment created with the environment_pathology.yml file.

Reviewer 2: yes, as long as you provided it first time you mention it.

> [Openslide](https://openslide.org/) is actually written in C, but you can also use Java or Python to work with is due to bindings.To run the code you will need to work in the `pathy` environment created with the environment_pathology.yml file.

Reviewer 2: we talk about openslide on line 63, where I suggest to include link.

```

> When we printed the properties we can see we are dealing with a tiled slide because we have properties like levels and tile height and width. These may not be the only properties we care about. We most likely may care about mapping this image back to real world sizes. We have a property that tells us the microns per pixel. Let's look:

Reviewer 2: What about other properties? Why do we have 9 levels?

Author: We could have more levels in a different image, but that is just the structure here. I'll try to clarify.

microns_per_pixel_y = slide_in.properties.get(openslide.PROPERTY_NAME_MPP_Y)

# convert to floats because they are strings
microns_per_pixel_x = float(microns_per_pixel_x)
microns_per_pixel_y = float(microns_per_pixel_y)

# calculate real-world dimensions of the slide in microns
real_world_width = width * microns_per_pixel_x
real_world_height = height * microns_per_pixel_y

# print the results
print(f"Real-world dimensions in microns: {real_world_width} microns (width), {real_world_height} microns (height)")
```

```output
Real-world dimensions in microns: 46000000.0 microns (width), 32914000.0 microns (height)
```

:::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::::::::::::::::

As you might guess loading this huge image is pointless if we just want a small image that gives us a general idea. We can use openslide to make a thumbnail.

```python
# get a thumbnail of the image and pop it out on your computer
slide_in_thumb_600 = slide_in.get_thumbnail(size=(600, 600))
slide_in_thumb_600.show() # 
```

With the above code an image will be opened on most operating systems, but if one does not open for you, do not depair.

> We could also store off our thumbnail (or any part of the image for that matter)as a numpy array, and then graph it.

Reviewer 2: ```suggestion
We could also store off our thumbnail (or any part of the image for that matter) as a numpy array, and then graph it.
```
